src/steadystateproblem.cpp

```diff
@@ -83,30 +84,67 @@ void SteadystateProblem::workSteadyStateProblem(Solver *solver, Model *model,
-                                                 NewtonSolver *newtonSolver,
-                                                 Model *model, int it) {
+void SteadystateProblem::findSteadyState(Solver *solver,
+                                         NewtonSolver *newtonSolver,
+                                         Model *model, int it) {
     /* First, try to run the Newton solver */
-    try {
-        applyNewtonsMethod(model, newtonSolver, NewtonStatus::newt);
-        if (maxSteps > 0)
-            std::copy_n(newtonSolver->getNumLinSteps().begin(),
-                        maxSteps, numlinsteps.begin());
-        return NewtonStatus::newt;
-    } catch (NewtonFailure const &) {
-        /* Newton solver failed. Get the number of steps used. */
-        if (maxSteps > 0)
-            std::copy_n(newtonSolver->getNumLinSteps().begin(),
-                        maxSteps, numlinsteps.begin());
-    }
+    findSteadyStateByNewtonsMethod(newtonSolver, model, false);
 
     /* Newton solver didn't work, so try to simulate to steady state */
+    if (!checkSteadyStateSuccess())
+        findSteadyStateBySimulation(solver, model, it);
+
+    /* Simulation didn't work, retry the Newton solver from last sim state. */
+    if (!checkSteadyStateSuccess())
+        findSteadyStateByNewtonsMethod(newtonSolver, model, true);
+
+    /* Nothing worked, throw an as informative error as possible */
+    if (!checkSteadyStateSuccess())
+        handleSteadyStateComputationFailure(solver, model);
+}
+
+void SteadystateProblem::findSteadyStateByNewtonsMethod(NewtonSolver *newtonSolver,
+                                                        Model *model,
+                                                        bool newton_retry) {
     try {
-        if (it < 1) /* No previous time point computed, set t = t0 */
-            t = model->t0();
-        else /* Carry on simulating from last point */
-            t = model->getTimepoint(it - 1);
+        applyNewtonsMethod(model, newtonSolver, newton_retry);
+        if (newton_retry)
+            steady_state_status[2] = SteadyStateStatus::success;
+        else
+            steady_state_status[0] = SteadyStateStatus::success;
+    } catch (NewtonFailure const &ex) {
+        /* nothing to be done */
+        switch (ex.error_code) {
+            case AMICI_TOO_MUCH_WORK:
+                steady_state_status[0] = SteadyStateStatus::failed_convergence;
+                break;
+            case AMICI_SINGULAR_JACOBIAN:
+                steady_state_status[0] = SteadyStateStatus::failed_factorization;
+                break;
+            case AMICI_DAMPING_FACTOR_ERROR:
+                steady_state_status[0] = SteadyStateStatus::failed_damping;
+                break;
+            default:
+                steady_state_status[0] = SteadyStateStatus::failed;
+        }
+    }
+
+    /* copy number of linear steps used */
+    if (maxSteps > 0)
+        std::copy_n(newtonSolver->getNumLinSteps().begin(),
```

Comment:
I believe this needs a check for `newton_retry` to switch between `numlinsteps.begin()` and `&numlinsteps.at(maxSteps)`

Reply:
True! Thanks!